### PR TITLE
SNOW-781049: support dot in table name

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-781049

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/session.py [hunk 1]
```diff
@@ -1974,10 +1974,14 @@ def _table_exists(self, table_name: str):
         # note: object name could have dots, e.g, a table could be created via: create table "abc.abc" (id int)
         # currently validate_object_name does not allow it, but if in the future we want to support the case, we need to
         # update the implementation accordingly in this method
-        qualified_table_name = table_name.split(".")
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/session.py:1977]: I think we allow `"db"."schema"."table"`, then will it break?

Comment (sfc-gh-aling) [reply src/snowflake/snowpark/session.py:1977]: I have updated the implementation, it should be able to handle all the special cases now

File: src/snowflake/snowpark/session.py [hunk 2]
```diff
@@ -166,6 +166,23 @@ def _remove_session(session: "Session") -> None:
             pass
 
 
+def _strip_double_quotes_in_like_statement_in_table_name(table_name: str) -> str:
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/session.py:169]: Can we move it to utils.py?

**Review by sfc-gh-jdu — COMMENTED** [review]
Can we fix the docstring of save_as_table and copy_into_table so it states clearly that when the input is a string, it only represents the table name; when the input is a list, it represents fully-qualified name?

File: src/snowflake/snowpark/session.py [hunk 3]
```diff
@@ -1980,41 +1997,45 @@ def query_history(self) -> QueryHistory:
...
+                # name in the form of "table"
+                tables = self._run_query(
+                    f"show tables like '{_strip_double_quotes_in_like_statement_in_table_name(qualified_table_name[0])}'"
+                )
+            elif len(qualified_table_name) == 2:
+                # name in the form of "schema.table" omitting database
+                # schema: qualified_table_name[0]
+                # table: qualified_table_name[1]
+                tables = self._run_query(
+                    f"show tables like '{_strip_double_quotes_in_like_statement_in_table_name(qualified_table_name[1])}' in schema {qualified_table_name[0]}"
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/session.py:2019]: Database and schema names are also identifiers. Do we need to do similar treatment as table?

File: src/snowflake/snowpark/session.py (hunk 2 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/session.py]: updated

File: src/snowflake/snowpark/_internal/analyzer/analyzer.py [hunk 4]
```diff
@@ -630,6 +630,7 @@ def do_resolve_with_resolved_children(
         if isinstance(logical_plan, SnowflakeCreateTable):
             return self.plan_builder.save_as_table(
                 logical_plan.table_name,
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/_internal/analyzer/analyzer.py]: updatd

File: src/snowflake/snowpark/session.py (hunk 3 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/session.py]: nope, `SHOW` command works like:
> SHOW tables like '<table_name>' in \<db\>.\<schema\>

<table_name> is expected to be a string constant as wrapped in single quotes `'`.

but db and schema after `in` can keep the double quotes as they are idenfitiers

**Review by sfc-gh-jdu — APPROVED** [review]

**Review by sfc-gh-sfan — APPROVED** [review]

**Pull request merged**